### PR TITLE
Add basic implementation of message call

### PR DESCRIPTION
**Pull request by hayesgm**
This patch adds a basic implementation of how message call will work in the EVM. We decide to use an abstract interface (not implemented yet) for `AccountInterface`, `ContractInterface` and `BlockInterface` that will allow the EVM to accept dependency injected interfaces for `get_account()`, `message_call()`, `get_block_header()` etc. This will give us a nice separation of concerns between the `Blockchain` project and the `EVM` project. We will use simple implementations of these to test the EVM opcode's implementation itself.

This PR needs to have test cases set-up and define what these interfaces will look like.

### CODE REVIEW

**Comment by masonforest** [comment]
I like it. Separation of concerns ftw!

**Comment by hayesgm** [comment]
Cool, this should be ready to go for a basic implementation of message call.

File: lib/block/header.ex [hunk 1]
```diff
@@ -0,0 +1,370 @@
...
+    number: integer() | nil,
+    gas_limit: EVM.val,
+    gas_used: EVM.val,
+    timestamp: EVM.timestamp | nil,
+    extra_data: binary(),
+    mix_hash: EVM.hash | nil,
+    nonce: <<_::64>> | nil, # TODO: 64-bit hash?
+  }
+
+  @d_0 131_072 # Eq.(40)
```
Comment (masonforest) [inline lib/block/header.ex]: Can we call this @homestead_block_number or something similar with a comment that it's `d0`?

**Comment by masonforest** [comment]
Ohhhhh, that's why the doctests weren't running! Thanks!

This looks great!

It looks like [the common tests for EnvironmentalInfo](https://github.com/ethereum/tests/tree/develop/VMTests/vmEnvironmentalInfo) are structured differently than the other tests :/ There's different file naming conventions and it looks like it's pulling in environment info from "filler" files. LMK if you want to hook those up this or if you want to hand them off ✋ 

Otherwise looks good to me 👍

**Comment by hayesgm** [comment]
There's a bunch of work to hook up to common tests and test this better than I am doing, but this is a good start. Going to merge now and we can true up. It's also nice to help understand the exact difference between `CALL`, `CALLCODE`, and `DELEGATECALL`.

**Pull request merged**